In [10]:
import pickle
import pandas as pd

In [11]:
# J'importe le fichier pkl

with open('dataframe.pkl', 'rb') as file:
    df = pickle.load(file)

In [12]:
df.describe(include='all')

No disposition Date mutation Nature mutation  Valeur fonciere  \
count     1.111078e+06       1111078         1111078     1.111078e+06   
unique             NaN           359               6              NaN   
top                NaN    30/04/2021           Vente              NaN   
freq               NaN          8006         1076794              NaN   
mean      1.051740e+00           NaN             NaN     3.178176e+05   
std       3.124035e-01           NaN             NaN     4.344339e+06   
min       1.000000e+00           NaN             NaN     1.000000e+02   
25%       1.000000e+00           NaN             NaN     4.800000e+04   
50%       1.000000e+00           NaN             NaN     1.300000e+05   
75%       1.000000e+00           NaN             NaN     2.360000e+05   
max       6.700000e+01           NaN             NaN     4.408400e+08   

       Type de voie        Voie   Code postal  Commune  Code commune  Section  \
count        704354     1108595  1.108609e+06  1111078  1.111078e+06  1111054   
unique          132      332246           NaN    30344           NaN      584   
top             RUE  LE VILLAGE           NaN     NICE           NaN        A   
freq         422712       14002           NaN     6340           NaN    78379   
mean            NaN         NaN  5.045689e+04      NaN  2.115873e+02      NaN   
std             NaN         NaN  2.695645e+04      NaN  1.679774e+02      NaN   
min             NaN         NaN  1.000000e+03      NaN  1.000000e+00      NaN   
25%             NaN         NaN  2.940000e+04      NaN  8.000000e+01      NaN   
50%             NaN         NaN  5.034000e+04      NaN  1.760000e+02      NaN   
75%             NaN         NaN  7.500600e+04      NaN  3.000000e+02      NaN   
max             NaN         NaN  9.749000e+04      NaN  9.090000e+02      NaN   

             No plan   1er lot  Surface Carrez du 1er lot  Nombre de lots  \
count   1.111078e+06  277300.0              108230.000000    1.111078e+06   
unique           NaN    7499.0                        NaN             NaN   
top              NaN       1.0                        NaN             NaN   
freq             NaN   10496.0                        NaN             NaN   
mean    4.240905e+02       NaN                  56.980472    3.182351e-01   
std     5.847717e+02       NaN                  85.567029    8.161554e-01   
min     1.000000e+00       NaN                   0.010000    0.000000e+00   
25%     9.300000e+01       NaN                  28.572500    0.000000e+00   
50%     2.330000e+02       NaN                  46.660000    0.000000e+00   
75%     5.180000e+02       NaN                  70.737500    0.000000e+00   
max     9.873000e+03       NaN                9614.000000    1.900000e+02   

       Type local  Surface reelle bati  Nombre pieces principales  \
count      656022        655035.000000              655035.000000   
unique          4                  NaN                        NaN   
top        Maison                  NaN                        NaN   
freq       366661                  NaN                        NaN   
mean          NaN            99.123564                   2.779458   
std           NaN           604.846448                   2.045638   
min           NaN             0.000000                   0.000000   
25%           NaN            33.000000                   1.000000   
50%           NaN            72.000000                   3.000000   
75%           NaN           100.000000                   4.000000   
max           NaN        145111.000000                  70.000000   

       Nature culture  Surface terrain  \
count          828285     8.282850e+05   
unique             27              NaN   
top                 S              NaN   
freq           484400              NaN   
mean              NaN     1.958955e+03   
std               NaN     8.613734e+03   
min               NaN     0.000000e+00   
25%               NaN     1.980000e+02   
50%               NaN     

In [14]:
pourcentage_manquant = (df.isna().sum() / len(df)) * 100
pourcentage_manquant

No disposition                0.000000
Date mutation                 0.000000
Nature mutation               0.000000
Valeur fonciere               0.000000
Type de voie                 36.606251
Voie                          0.223477
Code postal                   0.222217
Commune                       0.000000
Code commune                  0.000000
Section                       0.002160
No plan                       0.000000
1er lot                      75.042256
Surface Carrez du 1er lot    90.259010
Nombre de lots                0.000000
Type local                   40.956260
Surface reelle bati          41.045093
Nombre pieces principales    41.045093
Nature culture               25.452128
Surface terrain              25.452128
col_concat                    0.000000
dtype: float64

In [15]:
print(df.columns)

Index(['No disposition', 'Date mutation', 'Nature mutation', 'Valeur fonciere',
       'Type de voie', 'Voie', 'Code postal', 'Commune', 'Code commune',
       'Section', 'No plan', '1er lot', 'Surface Carrez du 1er lot',
       'Nombre de lots', 'Type local', 'Surface reelle bati',
       'Nombre pieces principales', 'Nature culture', 'Surface terrain',
       'col_concat'],
      dtype='object')


In [19]:
correlation = df[df.columns].corr()

sorted_correlation_Valeurfonciere = correlation['Valeur fonciere'].sort_values(ascending=False)

print(sorted_correlation_Valeurfonciere)

Valeur fonciere              1.000000
Surface Carrez du 1er lot    0.362928
Surface reelle bati          0.054003
Nombre pieces principales    0.021664
Code commune                 0.017872
No plan                      0.013262
Code postal                  0.009607
No disposition               0.007884
Surface terrain              0.001481
Nombre de lots              -0.005054
Name: Valeur fonciere, dtype: float64


# Arbre de décision pour predire le 'Type local'

In [38]:
# Sélectionner uniquement les lignes où 'Type local' est renseigné
df_filtered = df[df['Type local'].notna()]

In [57]:
# on prend quelques variables explicatives

X = df_filtered[["Surface reelle bati", "Nombre pieces principales"]]

#on nomme la variale cible

Y= df_filtered["Type local"]

In [58]:
#ok toutes les valeurs de Y sont renseignées
Y.isna().value_counts()

False    656022
Name: Type local, dtype: int64

In [59]:
Y.value_counts(normalize=True)

Maison                                      0.558916
Appartement                                 0.206357
Dépendance                                  0.165135
Local industriel. commercial ou assimilé    0.069592
Name: Type local, dtype: float64

In [113]:
# on scinde les données en echantillon test et train
#70% des données pour l’apprentissage
#30% des données pour l'échantillon test

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, Y,
                                                    test_size = 0.30,
                                                    stratify = Y,
                                                    random_state = 42)

On va gerer les valeurs manquantes dans l'echantillon d'apprentissage

In [119]:
X_test['Surface reelle bati'].fillna(X_test['Surface reelle bati'].mean(), inplace=True)

In [120]:
X_test['Surface reelle bati'].isna().value_counts()

False    196807
Name: Surface reelle bati, dtype: int64

In [121]:
X_train['Surface reelle bati'].fillna(X_test['Surface reelle bati'].mean(), inplace=True)

In [122]:
X_train['Surface reelle bati'].isna().value_counts()

False    459215
Name: Surface reelle bati, dtype: int64

In [123]:
X_test['Nombre pieces principales'].fillna(X_test['Surface reelle bati'].mean(), inplace=True)

In [124]:
X_test['Nombre pieces principales'].isna().value_counts()

False    196807
Name: Nombre pieces principales, dtype: int64

In [125]:
X_train['Nombre pieces principales'].fillna(X_test['Surface reelle bati'].mean(), inplace=True)

In [126]:
X_train['Nombre pieces principales'].isna().value_counts()

False    459215
Name: Nombre pieces principales, dtype: int64

# Apprentissage

In [127]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.tree import plot_tree

In [128]:
model_1 = DecisionTreeClassifier(max_depth=3 , min_samples_leaf=10, min_samples_split=20)
model_1 = model_1.fit(X_train,y_train)

# Prediction

In [131]:
y_pred = model_1.predict(X_test)
y_pred[0:10]

array(['Local industriel. commercial ou assimilé', 'Maison', 'Dépendance',
       'Maison', 'Dépendance', 'Appartement', 'Maison', 'Maison',
       'Maison', 'Dépendance'], dtype=object)

In [130]:
y_pred_proba = model_1.predict_proba(X_test)
y_pred_proba[0:10]

array([[5.55291131e-04, 0.00000000e+00, 9.95478344e-01, 3.96636522e-03],
       [6.60720043e-02, 5.37345513e-06, 3.68619022e-03, 9.30236432e-01],
       [1.29590752e-05, 9.82699635e-01, 1.72874064e-02, 0.00000000e+00],
       [2.75469392e-01, 0.00000000e+00, 0.00000000e+00, 7.24530608e-01],
       [1.29590752e-05, 9.82699635e-01, 1.72874064e-02, 0.00000000e+00],
       [7.54306156e-01, 0.00000000e+00, 0.00000000e+00, 2.45693844e-01],
       [2.75469392e-01, 0.00000000e+00, 0.00000000e+00, 7.24530608e-01],
       [6.60720043e-02, 5.37345513e-06, 3.68619022e-03, 9.30236432e-01],
       [2.75469392e-01, 0.00000000e+00, 0.00000000e+00, 7.24530608e-01],
       [1.29590752e-05, 9.82699635e-01, 1.72874064e-02, 0.00000000e+00]])

In [132]:
pd.crosstab(y_test,y_pred, colnames=['pred'], rownames=['obs'], margins=True)

pred                                      Appartement  Dépendance  \
obs                                                                 
Appartement                                     22623           0   
Dépendance                                          0       32500   
Local industriel. commercial ou assimilé            0         538   
Maison                                           7397           0   
All                                             30020       33038   

pred                                      Local industriel. commercial ou assimilé  \
obs                                                                                  
Appartement                                                                     74   
Dépendance                                                                       0   
Local industriel. commercial ou assimilé                                     12866   
Maison                                                                         165   
All                                                                          13105   

pred                                      Maison     All  
obs                                                       
Appartement                                17916   40613  
Dépendance                                     0   32500  
Local industriel. commercial ou assimilé     292   13696  
Maison                                    102436  109998  
All                                       120644  196807

In [135]:
from sklearn.metrics import recall_score, precision_score, f1_score

print('recall : ' + str(recall_score(y_test,y_pred,average='macro')))
print('precision : ' + str(precision_score(y_test,y_pred, average='macro')))
print('f1_score : ' + str(f1_score(y_test,y_pred, average='macro')))


recall : 0.856922511680977
precision : 0.8920381577537984
f1_score : 0.8701879038871843


# Optimisation des parametres de l'arbre

In [136]:
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import ShuffleSplit

In [137]:
parameters = {'max_depth' : [1,2,3,4,5,6] ,
              'min_samples_leaf' : [10,20,30,40],
              'min_samples_split' : [50,100,200]}

dtc = DecisionTreeClassifier()
dtc_model_1 = GridSearchCV(dtc, parameters, scoring = 'accuracy' ,
                   cv = ShuffleSplit(test_size=0.30, n_splits=1, random_state=0))
dtc_model_1.fit(X_train, y_train )

print("Voici les paramètres du meilleure modèle : " + str(dtc_model_1.best_estimator_))
print("Voici le score "  + str(dtc_model_1.scorer_) + " du meilleure modèle : " + str(dtc_model_1.best_score_))

Voici les paramètres du meilleure modèle : DecisionTreeClassifier(max_depth=6, min_samples_leaf=10, min_samples_split=100)
Voici le score make_scorer(accuracy_score) du meilleure modèle : 0.8700831125467281


Comment c'est possible que avec le gridsearch le score make_scorer(accuracy_score) du meilleure modèle : 0.8700831125467281 et que sans le gridsearch precision : 0.8920381577537984??